_P221_LandisGyrUH550.ino

```diff
@@ -0,0 +1,450 @@
+                  addLog(LOG_LEVEL_ERROR, log);
+                  break;
+                }
+                lineComplete = false;
+                bufpos = 0;
+              }  // if (lineComplete && initComplete)
+            } // while r != 0x21
+            
+            messageProcessingComplete = true;
+
+            String log = F("LUGUH: Datasets found: ");
+            log += indexDS;
+            Serial.println(log);
+
+            for (byte i = 0; i<indexDS; i++)
+            {
+              String log = F("LUGUH: <DataSet> OBIS: ");
+              log += datasets[i].address;
+              for (byte j = 0; j < datasets[i].vupCount; j++)
+              {
+                log += F(" value: ");
+                log += datasets[i].vup[j].value;
+
+                if (datasets[i].vup[j].unit != NULL)
+                {
+                  log += F(" unit: ");
+                  log += datasets[i].vup[j].unit;
+                }
+              }
+              addLog(LOG_LEVEL_INFO, log);
+              Serial.println(log);
+
+            }
+
+            // fill UserVars ...
+            UserVar[event->BaseVarIndex] = readFromDataSet("6.8", 1, indexDS);
+
+            lineComplete = false;
+            initComplete = false;
+            break;
+          }
+          case 1:
+          {
+            UserVar[event->BaseVarIndex] = readFromDataSet("6.26", 1, indexDS);
+            break;
+          }
+          case 2:
+          {
+            UserVar[event->BaseVarIndex] = readFromDataSet("6.8*01", 1, indexDS);
+            break;
+          }
+          case 3:
+          {
+            UserVar[event->BaseVarIndex] = readFromDataSet("6.26*01", 1, indexDS);
+            break;
+          }
+          case 4:
+          {
+            UserVar[event->BaseVarIndex] = readFromDataSet("6.6", 1, indexDS);
+            break;
+          }
+          case 5:
+          {
+            UserVar[event->BaseVarIndex] = readFromDataSet("9.4", 1, indexDS);
+            break;
+          }
+          case 6:
+          {
+            UserVar[event->BaseVarIndex] = readFromDataSet("9.4", 2, indexDS);
+            break;
+          }
+        }
+      } // if (runTask)... 
+      success = true;
+      break;
+    } // case PLUGIN_READ
+  }
+  return success;
+}
+
+bool parityCheck(unsigned input)
+{
+  bool inputparity = input & 128;
+  int x = input & 127;
+
+  int parity = 0;
+  while (x != 0)
+  {
+    parity ^= x;
+    x >>= 1;
+  }
+
+  if ((parity & 0x1) != inputparity)
+    return (1);
+  else
+    return (0);
+}
+
+void initDataSet()
+{
+  indexDS = 0;
+  for(byte i = 0; i < 10; i++)
+  {
+    datasets[i].address[0] = '\0';
+    datasets[i].vupCount = 0;
+    for (byte j = 0; j < 20; j++)
+    {
+      datasets[i].vup[j].value[0] = '\0';
+      datasets[i].vup[j].unit[0] = '\0';
+    }
+  }
+}
+
+float readFromDataSet(const char OBIS[], byte valueToRead, byte dataSetSize)
+{
+  while (!messageProcessingComplete)
+  {
+    delay(500);

```

Comment:
Do not call delay, it is blocking all other tasks on the node.
Especially these big delays running in a while loop will cause all kinds of issues.
For example disconnects from  MQTT broker, timeout on web loading, no response to background tasks, missing requests leading to missing ARP packages and thus the node being unreachable from the network.
And very likely some buffers will overflow and leading to crashes.

If you need to wait, reschedule a call to see when you think it is ready, or try reading from the 10/sec call routine.
See the code in BME280 plugin to get some suggestions.

Reply:
the BME280 plugin was not of so much help, but I plan to implement asynchronous tasks for meter reading and value publishing with reading done in normal read cycle and publishing only done when new values are in memory, triggered once a second. Doing so will completely avoid any delays.

Reply:
eleminated delay()